core/finality-grandpa/src/communication/gossip.rs

```diff
@@ -0,0 +1,807 @@
+		super::global_topic::<B>(s.0),
+		super::round_topic::<B>(view.round.0, s.0),
+	];
+
+	if view.round.0 != 0 {
+		let r = Round(view.round.0 - 1);
+		topics.push(super::round_topic::<B>(r.0, s.0))
+	}
+
+	topics
+}
+
+/// Grandpa gossip message type.
+/// This is the root type that gets encoded and sent on the network.
+#[derive(Debug, Encode, Decode)]
+pub(super) enum GossipMessage<Block: BlockT> {
+	/// Grandpa message with round and set info.
+	VoteOrPrecommit(VoteOrPrecommitMessage<Block>),
+	/// Grandpa commit message with round and set info.
+	Commit(FullCommitMessage<Block>),
+	/// A neighbor packet. Not repropagated.
+	Neighbor(VersionedNeighborPacket<NumberFor<Block>>),
+}
+
+/// Network level message with topic information.
+#[derive(Debug, Encode, Decode)]
+pub(super) struct VoteOrPrecommitMessage<Block: BlockT> {
+	/// The round this message is from.
+	pub(super) round: Round,
+	/// The voter set ID this message is from.
+	pub(super) set_id: SetId,
+	/// The message itself.
+	pub(super) message: SignedMessage<Block>,
+}
+
+/// Network level commit message with topic information.
+#[derive(Debug, Encode, Decode)]
+pub(super) struct FullCommitMessage<Block: BlockT> {
+	/// The round this message is from.
+	pub(super) round: Round,
+	/// The voter set ID this message is from.
+	pub(super) set_id: SetId,
+	/// The compact commit message.
+	pub(super) message: CompactCommit<Block>,
+}
+
+/// V1 neighbor packet. Neighbor packets are sent from nodes to their peers
+/// and are not repropagated. These contain information about the node's state.
+#[derive(Debug, Encode, Decode)]
+pub(super) struct NeighborPacket<N> {
+	round: Round,
+	set_id: SetId,
+	commit_finalized_height: N,
+}
+
+/// A versioned neighbor packet.
+#[derive(Debug, Encode, Decode)]
+pub(super) enum VersionedNeighborPacket<N> {
+	#[codec(index = "1")]
+	V1(NeighborPacket<N>),
+}
+
+impl<N> VersionedNeighborPacket<N> {
+	fn into_neighbor_packet(self) -> NeighborPacket<N> {
+		match self {
+			VersionedNeighborPacket::V1(p) => p,
+		}
+	}
+}
+
+/// Misbehavior that peers can perform.
+///
+/// `cost` gives a cost that can be used to perform cost/benefit analysis of a
+/// peer.
+#[derive(Clone, Copy, Debug)]
+enum Misbehavior {
+	// invalid neighbor message, considering the last one.
+	InvalidViewChange,
+	// could not decode neighbor message. bytes-length of the packet.
+	UndecodablePacket(i32),
+	// Bad commit message
+	BadCommitMessage {
+		signatures_checked: i32,
+		blocks_loaded: i32,
+		equivocations_caught: i32,
+	},
+	// A message received that's from the future relative to our view.
+	// always misbehavior.
+	FutureMessage,
+}
+
+impl Misbehavior {
+	fn cost(&self) -> i32 {
+		use Misbehavior::*;
+
+		match *self {
+			InvalidViewChange => -500,
+			UndecodablePacket(bytes) =>  bytes.saturating_mul(-5),
+			BadCommitMessage { signatures_checked, blocks_loaded, equivocations_caught } => {
+				let cost = 25i32.saturating_mul(signatures_checked)
```

Comment:
Maybe make these numbers to constant variables.

Reply:
Agreed.  Also, there needs to be documentation that explains why these values were chose.

Reply:
@DemiMarie-parity they're completely arbitrary for now, tweaking can happen in a follow-up. This is more about the ability to punish at least somewhat for certain misbehaviors, which is an improvement over the current state.